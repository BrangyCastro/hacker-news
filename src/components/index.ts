export { Header } from "./Header/Header";
export { Tabs } from "./Tabs/Tabs";
export { Card } from "./Card/Card";
export { Dropdown } from "./Dropdown/Dropdown";
