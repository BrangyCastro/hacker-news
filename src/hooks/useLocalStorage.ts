import { useState, useEffect } from "react";

export function useLocalStorage<T>(key: string, defaultValue?: T) {
  const getStorageValue = () => {
    const item = localStorage.getItem(key);
    return item ? JSON.parse(item) : defaultValue;
  };

  const [storedValue, setStoredValue] = useState(getStorageValue);

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(storedValue));
  }, [key, storedValue]);

  const setValue = (value: T) => {
    const newValue = value instanceof Function ? value(storedValue) : value;
    localStorage.setItem(key, JSON.stringify(newValue));
    setStoredValue(newValue);
  };

  const findValue = (id: number) => {
    const removeFave = storedValue.find((fave: any) => fave.story_id === id);
    if (!!removeFave) {
      return true;
    }
    return false;
  };

  const removeValue = (id: number) => {
    const oldData = getStorageValue();
    const removeFave = oldData.find((fave: any) => fave.story_id === id);

    let index = oldData.indexOf(removeFave);
    oldData.splice(index, 1);
    setValue(oldData);
  };

  return [storedValue, setValue, findValue, removeValue];
}
