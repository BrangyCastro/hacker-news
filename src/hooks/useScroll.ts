import { useEffect, useState } from "react";
import { searchNews } from "../api/news";

interface Props {
  filter: string;
  page: number;
  setData: React.Dispatch<
    React.SetStateAction<{
      news: never[];
      page: number;
      loadingFilter: boolean;
    }>
  >;
}

export const useScroll = ({ filter, page, setData }: Props) => {
  const [isFetching, setIsFetching] = useState(false);

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  useEffect(() => {
    if (!isFetching) return;
    fetchMoreNews();
  }, [isFetching]);

  const handleScroll = () => {
    if (
      window.innerHeight + document.documentElement.scrollTop !==
        document.documentElement.offsetHeight ||
      isFetching
    )
      return;
    setIsFetching(true);
  };

  const fetchMoreNews = async () => {
    const resp = await searchNews(filter, page);
    setData((old) => ({
      ...old,
      news: old.news.concat(resp.news),
      page: resp.page,
    }));
    setIsFetching(false);
  };

  return [isFetching] as const;
};
