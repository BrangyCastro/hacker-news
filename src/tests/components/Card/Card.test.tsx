import React from "react";
import { shallow } from "enzyme";
import { renderHook, act } from "@testing-library/react-hooks";
import moment from "moment";
import { Card } from "../../../components";
import { dataNews } from "../../fixtures/demoNews";
import { useLocalStorage } from "../../../hooks/useLocalStorage";

describe("Card tests", () => {
  const news = {
    story_id: 4,
    created_at: "2021-12-01T20:54:12.000Z",
    author: "brangycastro",
    story_title: "Test notice",
    story_url: "hjadshjsd",
  };

  const deleteFavorite = jest.fn();

  const wrapper = shallow(
    <Card news={news} value={dataNews} deleteFavorite={deleteFavorite} />
  );

  test("should be displayed correctly", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".card").exists()).toBe(true);
  });

  test("should render the new news", () => {
    const title = wrapper.find("h1").text().trim();
    expect(title).toBe(news.story_title);
    const span = wrapper.find("span").text().trim();
    expect(span).toBe(
      `${moment(news.created_at).startOf("hour").fromNow()} by ${news.author}`
    );
  });

  test("should add a news to favorite", () => {
    const wrapper = shallow(<Card news={news} value={dataNews} />);
    wrapper.find("#add").simulate("click");
    const { result } = renderHook(() => useLocalStorage("fave"));
    const [, , find] = result.current;
    let findValue;
    act(() => {
      findValue = find(news.story_id);
    });
    expect(findValue).toBeTruthy();
  });

  test("should delete a news to favorite", () => {
    wrapper.find("#delete").simulate("click");
    const { result } = renderHook(() => useLocalStorage("fave"));
    const [, , , remove] = result.current;
    act(() => {
      remove(news.story_id);
    });
    const storage = JSON.parse(localStorage.getItem("fave") || "[]");
    expect(storage.length).toBe(dataNews.length - 1);
  });
});
