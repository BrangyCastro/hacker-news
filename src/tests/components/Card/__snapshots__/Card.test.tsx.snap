// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card tests should be displayed correctly 1`] = `
<div
  className="card"
>
  <div
    className="card-body"
  >
    <a
      href="hjadshjsd"
      rel="noreferrer"
      target="_blank"
    >
      <div
        className="time"
      >
        <img
          alt=""
          src="time.png"
        />
        <span>
          4 days ago
           by 
          brangycastro
        </span>
      </div>
      <h1>
        Test notice
      </h1>
    </a>
  </div>
  <div
    className="card-button"
  >
    <img
      alt=""
      id="delete"
      onClick={[Function]}
      src="favorite.png"
    />
  </div>
</div>
`;
